x-pack/apm-server/aggregation/lsm.go

```diff
@@ -0,0 +1,132 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package aggregation
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.uber.org/zap"
+
+	"github.com/elastic/apm-aggregation/aggregationpb"
+	"github.com/elastic/apm-aggregation/aggregators"
+	"github.com/elastic/apm-data/model/modelpb"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type Aggregator struct {
+	logger         *zap.Logger
+	baseaggregator *aggregators.Aggregator
+}
+
+// NewAggregator returns a new instance of aggregator.
+func NewAggregator(
+	ctx context.Context,
+	maxSvcs, maxTxGroups, maxSvcTxGroups, maxSpanGroups int,
+	nextProcessor modelpb.BatchProcessor,
+	logger *logp.Logger,
+) (*Aggregator, error) {
+	zapLogger := zap.New(logger.Core(), zap.WithCaller(true)).Named("aggregator")
+
+	baseaggregator, err := aggregators.New(
+		aggregators.WithLimits(aggregators.Limits{
+			MaxSpanGroups:                         maxSpanGroups,
+			MaxSpanGroupsPerService:               max(maxSpanGroups/10, 1),
+			MaxTransactionGroups:                  maxTxGroups,
+			MaxTransactionGroupsPerService:        max(maxTxGroups/10, 1),
+			MaxServiceTransactionGroups:           maxSvcTxGroups,
+			MaxServiceTransactionGroupsPerService: max(maxSvcTxGroups/10, 1),
+			MaxServiceInstanceGroupsPerService:    max(maxSvcs/10, 1),
```

Comment:
We have not yet settled on whether to keep service instance or to revert it. In any case, it should not be blocking this PR. Meanwhile, is `maxSvcs/10` a sane enough default for MaxServiceInstanceGroupsPerService?

Reply:
I think it's fine to come back to it in a followup, but soon. Let's go with `maxSvcs/10` for now and create an issue to finalise the decision.